pkg/mutation/mutation_function.go

```diff
@@ -0,0 +1,121 @@
+package mutation
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func Mutate(mutator MutatorWithSchema, obj *unstructured.Unstructured) error {
+	return mutate(mutator, obj.Object, nil, 0)
+}
+
+func mutate(m MutatorWithSchema, current interface{}, previous interface{}, depth int) error {
+	if len(m.Path().Nodes)-1 == depth {
+		return addValue(m, current, previous, depth)
+	}
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		next, ok := current.(map[string]interface{})[pathEntry.(*parser.Object).Reference]
+		if !ok {
+			next = createMissingElement(m, current, previous, depth)
+		}
+		if err := mutate(m, next, current, depth+1); err != nil {
+			return err
+		}
+		return nil
+	case parser.ListNode:
+		elementFound := false
+		listPathEntry := pathEntry.(*parser.List)
+		glob := listPathEntry.Glob
+		key := listPathEntry.KeyField
+		for _, listElement := range current.([]interface{}) {
+			if glob {
+				if err := mutate(m, listElement, current, depth+1); err != nil {
+					return err
+				}
+				elementFound = true
+			} else if elementValue, ok := listElement.(map[string]interface{})[key]; ok {
+				if *listPathEntry.KeyValue == elementValue {
+					if err := mutate(m, listElement, current, depth+1); err != nil {
+						return err
+					}
+					elementFound = true
+				}
+			}
+		}
+		// If no matching element in the array was found in non Globbed list, create a new element
+		if !listPathEntry.Glob && !elementFound {
+			next := createMissingElement(m, current, previous, depth)
+			if err := mutate(m, next, current, depth+1); err != nil {
+				return err
+			}
+		}
+	default:
+		return fmt.Errorf("Unrecognized type: %v", t)
+	}
+	return nil
+}
+
+func addValue(m MutatorWithSchema, current interface{}, previous interface{}, depth int) error {
+	// TODO: it should be considered if the value set can be not just a simple string, but json which could be unmarshalled into an object
```

Comment:
Raw won't ever be anything else than an object.
I think here https://github.com/open-policy-agent/gatekeeper/pull/916#discussion_r515445603 we agreed to use assign.value (and in https://github.com/open-policy-agent/gatekeeper/pull/978 we make sure the field is there).
The good thing is, if we do that we'll just assign what's there and it will work, regardless of it being a string or an object.

Reply:
+1, though there is the added complexity that label values are always strongly typed as `string`, so we can't just set an `interface{}` like we can for generic mutations.

Reply:
We _could_ trust the validation here https://github.com/open-policy-agent/gatekeeper/pull/978/files#diff-763d73379a260d55f660904fdb401a3582a47d68e3308400b230d47d6afe82dfR148 that checks for value to be a string during insertion, and assert / panic here.
Wdyt?

Reply:
It might make it so one invalid mutation makes the rest of the mutation webhook go down. The only way I could justify to myself that it was okay to do so in the `schema` code is because if the schema is corrupt we are in a truly unknown state and trying to rebuild in-process would leave us with a webhook serving unknown state during the rebuilding period (or random timeouts if we retain the lock).

what if we add a nullable `valueValidator` parameter that can return an error if the value is in an incorrect form? If it returns a non-nil error, we bubble it up. That way the mutation function remains agnostic.

Slightly better would be to make a `valueMutator` struct that can be instantiated with these pluggable checks to simplify argument passing.

Reply:
I insist on validating the object :-)
There's nothing dynamic in the object itself, so either Value() returns a a string or not. The check does not have to happen that late.

If we validate the object in the controller too (and discard it if it does not fit), there's no chance value is not a string when it gets into the cache (unless bugs, of course).

Reply:
I'm a bit confused? I thought we both were for validating the object?

A failed, unchecked typecast will lead to a `panic()`, I am against panics in this part of the code.

Reply:
What I was saying is that if we do the validation in the controller it will be impossible to have assignmetadatas which value is not a string.
But agree on the extra paranoia, let's go defensive.

Reply:
Both checks (unmarshaling and ==string) are done on Mutator creation, so skipping them here.

Reply:
Plus, `Value()` will return an error for `AssignMetadataMutator` if the value is not a string.
I think this covers what we want

Reply:
I had forgotten that we are adding these to unstructured anyway, so there is no type casting needed. This works.